lib/usd/translators/shading/mtlxBaseWriter.cpp

```diff
@@ -190,63 +157,112 @@ UsdAttribute MtlxUsd_BaseWriter::AddSwizzle(
     }
 
     // The swizzle name varies according to source and destination channel sizes:
-    std::string srcType, dstType;
-
-    switch (numChannels) {
-    case 1:
-        srcType = "float";
-        swizzleSchema.CreateInput(TrMtlxTokens->in, SdfValueTypeNames->Float)
-            .ConnectToSource(UsdShadeOutput(nodeOutput));
-        break;
-    case 2:
-        srcType = "vector2";
-        swizzleSchema.CreateInput(TrMtlxTokens->in, SdfValueTypeNames->Float2)
-            .ConnectToSource(UsdShadeOutput(nodeOutput));
-        break;
-    case 3:
-        srcType = "color3";
-        swizzleSchema.CreateInput(TrMtlxTokens->in, SdfValueTypeNames->Color3f)
-            .ConnectToSource(UsdShadeOutput(nodeOutput));
-        break;
-    case 4:
-        srcType = "color4";
-        swizzleSchema.CreateInput(TrMtlxTokens->in, SdfValueTypeNames->Color4f)
-            .ConnectToSource(UsdShadeOutput(nodeOutput));
-        break;
-    default: TF_CODING_ERROR("Unsupported format for swizzle"); return UsdAttribute();
-    }
+    swizzleSchema.CreateInput(TrMtlxTokens->in, sourceType)
+        .ConnectToSource(UsdShadeOutput(nodeOutput));
 
     swizzleSchema.CreateInput(TrMtlxTokens->channels, SdfValueTypeNames->String)
-        .Set(channels, UsdTimeCode::Default());
+        .Set(channel, UsdTimeCode::Default());
 
-    switch (channels.size()) {
-    case 1:
-        dstType = "float";
-        swizzleOutput = swizzleSchema.CreateOutput(TrMtlxTokens->out, SdfValueTypeNames->Float);
-        break;
-    case 2:
-        dstType = "vector2";
-        swizzleOutput = swizzleSchema.CreateOutput(TrMtlxTokens->out, SdfValueTypeNames->Float2);
-        break;
-    case 3:
-        dstType = "color3";
-        swizzleOutput = swizzleSchema.CreateOutput(TrMtlxTokens->out, SdfValueTypeNames->Color3f);
-        break;
-    case 4:
-        dstType = "color4";
-        swizzleOutput = swizzleSchema.CreateOutput(TrMtlxTokens->out, SdfValueTypeNames->Color4f);
-        break;
-    }
+    swizzleOutput = swizzleSchema.CreateOutput(TrMtlxTokens->out, SdfValueTypeNames->Float);
 
-    TfToken swizzleID(TfStringPrintf("ND_swizzle_%s_%s", srcType.c_str(), dstType.c_str()));
+    TfToken swizzleID(TfStringPrintf("ND_swizzle_%s_float", materialXType.c_str()));
 
     swizzleSchema.CreateIdAttr(VtValue(swizzleID));
 
     return swizzleOutput;
 }
 
 UsdAttribute
-MtlxUsd_BaseWriter::AddSwizzleConversion(const SdfValueTypeName& destType, UsdAttribute nodeOutput)
+MtlxUsd_BaseWriter::AddConstructor(UsdAttribute nodeInput, size_t channelIndex, MPlug inputPlug)
+{
```

Comment:
A constructor is a node with one float input per component of a color[34] or vector[234] type. This allows connecting to the R of a color or the Y of a vector2 input.